File: apps/api/src/services/admin.service.ts
import prisma from "../lib/prisma";
import { AppError } from "../middleware/errorHandler";

export async function getMinistryStats(): Promise<any> {
  // Supabase pooler exposes a single connection (connection_limit=1), so queries
  // MUST run sequentially; parallel Promise.all calls contend for one connection
  // and time out with Prisma P2024.
  const run = async <T>(fn: () => Promise<T>): Promise<T> => fn();
  const coopCount = await run(() => prisma.coOp.count({}));
  const workerCount = await run(() => prisma.workerProfile.count({}));
  const pendingWorkers = await run(() =>
    prisma.workerProfile.count({ where: { status: "PENDING_ADMIN_APPROVAL" } })
  );
  const consumerCount = await run(() =>
    prisma.user.count({ where: { role: "CONSUMER" } })
  );
  const bookingCount = await run(() => prisma.booking.count({}));
  const revenue = await run(() =>
    prisma.booking.aggregate({ _sum: { commissionAmount: true } })
  );
  const verifiedWorkerCount = await run(() =>
    prisma.workerProfile.count({ where: { status: "VERIFIED" } })
  );
  const suspendedWorkerCount = await run(() =>
    prisma.workerProfile.count({ where: { status: "SUSPENDED" } })
  );

  return {
    totalCoops: coopCount,
    totalWorkers: workerCount,
    verifiedWorkers: verifiedWorkerCount,
    pendingWorkers,
    suspendedWorkers: suspendedWorkerCount,
    totalConsumers: consumerCount,
    totalBookings: bookingCount,
    platformRevenue: Number(revenue._sum.commissionAmount || 0),
  };
}

export async function listCoops(): Promise<any[]> {
  const coops = await prisma.coOp.findMany({
    include: {
      _count: { select: { workers: true, services: true } },
      admin: { select: { id: true, user: { select: { id: true, name: true, phone: true } } } },
    },
    orderBy: { createdAt: "asc" },
  });

  // Sequential to avoid P2024 under Supabase's single-connection pooler.
  const result: any[] = [];
  for (const c of coops) {
    const bookingStats = await prisma.booking.aggregate({
      where: { service: { coopId: c.id } },
      _count: true,
      _sum: { commissionAmount: true },
    });
    result.push({
      id: c.id, name: c.name, registrationNo: c.registrationNo, description: c.description,
      address: c.address, city: c.city, state: c.state, pincode: c.pincode,
      latitude: c.latitude, longitude: c.longitude, radiusKm: c.radiusKm,
      commissionRate: Number(c.commissionRate), isActive: c.isActive, createdAt: c.createdAt,
      admin: c.admin?.user ? { id: c.admin.user.id, name: c.admin.user.name, phone: c.admin.user.phone } : null,
      workerCount: c._count.workers,
      serviceCount: c._count.services,
      revenue: Number(bookingStats._sum.commissionAmount || 0),
    });
  }
  return result;
}

export async function listAllWorkers(status?: string, q?: string): Promise<any[]> {
  const where: any = {};
  if (status) {
    if (status === "ALL") {
      // no filter
    } else {
      where.status = status;
    }
  }
  if (q) {
    where.OR = [
      { user: { name: { contains: q, mode: "insensitive" } } },
      { user: { phone: { contains: q } } },
      { workAddress: { contains: q, mode: "insensitive" } },
    ];
  }

  const profiles = await prisma.workerProfile.findMany({
    where,
    include: {
      user: { select: { id: true, name: true, phone: true, avatarUrl: true, createdAt: true } },
      coop: { select: { id: true, name: true, city: true } },
    },
    orderBy: { createdAt: "desc" },
  });

  return profiles.map((p) => ({
    id: p.id,
    userId: p.userId,
    name: p.user.name,
    phone: p.user.phone,
    avatarUrl: p.user.avatarUrl,
    joinedAt: p.user.createdAt,
    status: p.status,
    skillTags: p.skillTags,
    bio: p.bio,
    workAddress: p.workAddress,
    latitude: p.latitude,
    longitude: p.longitude,
    phoneVerified: p.phoneVerified,
    kycStatus: p.kycStatus,
    kycDocumentUrl: p.kycDocumentUrl,
    aadhaarVerified: p.aadhaarVerified,
    avgRating: p.avgRating,
    totalJobs: p.totalJobs,
    coop: p.coop ? { id: p.coop.id, name: p.coop.name, city: p.coop.city } : null,
  }));
}

export async function approveWorkerByAdmin(workerId: string, note?: string): Promise<any> {
  const profile = await prisma.workerProfile.findUnique({ where: { id: workerId } });
  if (!profile) throw new AppError("Worker profile not found", 404);
  const updated = await prisma.workerProfile.update({
    where: { id: workerId },
    data: { status: "VERIFIED", kycStatus: "VERIFIED", aadhaarVerified: true },
    include: { user: { select: { id: true, name: true } } },
  });
  return {
    id: updated.id,
    name: updated.user.name,
    status: updated.status,
  };
}

export async function rejectWorkerByAdmin(workerId: string, reason: string): Promise<any> {
  const profile = await prisma.workerProfile.findUnique({ where: { id: workerId } });
  if (!profile) throw new AppError("Worker profile not found", 404);
  const updated = await prisma.workerProfile.update({
    where: { id: workerId },
    data: { status: "SUSPENDED", kycStatus: "REJECTED" },
    include: { user: { select: { id: true, name: true } } },
  });
  return {
    id: updated.id,
    name: updated.user.name,
    status: updated.status,
  };
}

File: backend/app/services/admin_service.py
from ..db import db
from ..errors import AppError
from ..utils import num, deep_serialize, now_utc


async def get_ministry_stats() -> dict:
    coop_count = await db.fetchval('SELECT COUNT(*) FROM "CoOp"')
    worker_count = await db.fetchval('SELECT COUNT(*) FROM "WorkerProfile"')
    pending_workers = await db.fetchval(
        'SELECT COUNT(*) FROM "WorkerProfile" WHERE status=$1', "PENDING_ADMIN_APPROVAL"
    )
    consumer_count = await db.fetchval(
        'SELECT COUNT(*) FROM "User" WHERE role=$1', "CONSUMER"
    )
    booking_count = await db.fetchval('SELECT COUNT(*) FROM "Booking"')
    revenue = await db.fetchrow(
        'SELECT COALESCE(SUM("commissionAmount"),0) AS total FROM "Booking"'
    )
    verified_workers = await db.fetchval(
        'SELECT COUNT(*) FROM "WorkerProfile" WHERE status=$1', "VERIFIED"
    )
    suspended_workers = await db.fetchval(
        'SELECT COUNT(*) FROM "WorkerProfile" WHERE status=$1', "SUSPENDED"
    )
    return {
        "totalCoops": coop_count,
        "totalWorkers": worker_count,
        "verifiedWorkers": verified_workers,
        "pendingWorkers": pending_workers,
        "suspendedWorkers": suspended_workers,
        "totalConsumers": consumer_count,
        "totalBookings": booking_count,
        "platformRevenue": num(revenue["total"]) if revenue is not None else 0,
    }


async def list_coops() -> list[dict]:
    coops = await db.fetch('SELECT * FROM "CoOp" ORDER BY "createdAt" ASC')
    result = []
    for c in coops:
        admin_row = await db.fetchrow(
            'SELECT u.id, u.name, u.phone FROM "CoopAdminProfile" cap '
            'JOIN "User" u ON u.id=cap."userId" WHERE cap."coopId"=$1',
            c["id"],
        )
        worker_count = await db.fetchval('SELECT COUNT(*) FROM "WorkerProfile" WHERE "coopId"=$1', c["id"])
        service_count = await db.fetchval('SELECT COUNT(*) FROM "Service" WHERE "coopId"=$1', c["id"])
        revenue_row = await db.fetchrow(
            'SELECT COALESCE(SUM(b."commissionAmount"),0) AS ca FROM "Booking" b '
            'JOIN "Service" s ON s.id=b."serviceId" WHERE s."coopId"=$1',
            c["id"],
        )
        result.append({
            "id": c["id"],
            "name": c["name"],
            "registrationNo": c["registrationNo"],
            "description": c["description"],
            "address": c["address"],
            "city": c["city"],
            "state": c["state"],
            "pincode": c["pincode"],
            "latitude": c["latitude"],
            "longitude": c["longitude"],
            "radiusKm": c["radiusKm"],
            "commissionRate": num(c["commissionRate"]),
            "isActive": c["isActive"],
            "createdAt": c["createdAt"],
            "admin": admin_row,
            "workerCount": worker_count,
            "serviceCount": service_count,
            "revenue": num(revenue_row["ca"]) if revenue_row is not None else 0,
        })
    return deep_serialize(result)


async def list_all_workers(status: str | None = None, q: str | None = None) -> list[dict]:
    conditions = []
    params = []
    if status and status != "ALL":
        params.append(status)
        conditions.append(f"wp.status=${len(params)}")
    if q:
        params.append(f"%{q}%")
        conditions.append(
            f"(u.name ILIKE ${len(params)} OR u.phone ILIKE ${len(params)} "
            f"OR wp.\"workAddress\" ILIKE ${len(params)})"
        )
    where = (" WHERE " + " AND ".join(conditions)) if conditions else ""
    rows = await db.fetch(
        f"""
        SELECT wp.*, u.id AS "user_id", u.name AS "user_name", u.phone AS "user_phone",
               u."avatarUrl" AS "user_avatarUrl", u."createdAt" AS "user_createdAt",
               c.id AS "coop_id", c.name AS "coop_name", c.city AS "coop_city"
        FROM "WorkerProfile" wp
        JOIN "User" u ON u.id=wp."userId"
        LEFT JOIN "CoOp" c ON c.id=wp."coopId"
        {where}
        ORDER BY wp."createdAt" DESC
        """,
        *params,
    )
    result = []
    for w in rows:
        result.append({
            "id": w["id"],
            "userId": w["userId"],
            "name": w["user_name"],
            "phone": w["user_phone"],
            "avatarUrl": w["user_avatarUrl"],
            "joinedAt": w["user_createdAt"],
            "status": w["status"],
            "skillTags": w["skillTags"],
            "bio": w["bio"],
            "workAddress": w["workAddress"],
            "latitude": w["latitude"],
            "longitude": w["longitude"],
            "phoneVerified": w["phoneVerified"],
            "kycStatus": w["kycStatus"],
            "kycDocumentUrl": w["kycDocumentUrl"],
            "aadhaarVerified": w["aadhaarVerified"],
            "avgRating": w["avgRating"],
            "totalJobs": w["totalJobs"],
            "coop": (
                {"id": w["coop_id"], "name": w["coop_name"], "city": w["coop_city"]}
                if w["coop_id"] else None
            ),
        })
    return deep_serialize(result)


async def approve_worker_by_admin(worker_id: str, note: str | None = None) -> dict:
    profile = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if profile is None:
        raise AppError("Worker profile not found", 404)
    await db.execute(
        'UPDATE "WorkerProfile" SET status=$1, "kycStatus"=$2, "aadhaarVerified"=true, "updatedAt"=$3 WHERE id=$4',
        "VERIFIED", "VERIFIED", now_utc(), worker_id,
    )
    user = await db.fetchrow('SELECT id, name FROM "User" WHERE id=$1', profile["userId"])
    return {"id": worker_id, "name": user["name"] if user else None, "status": "VERIFIED"}


async def reject_worker_by_admin(worker_id: str, reason: str) -> dict:
    profile = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if profile is None:
        raise AppError("Worker profile not found", 404)
    await db.execute(
        'UPDATE "WorkerProfile" SET status=$1, "kycStatus"=$2, "updatedAt"=$3 WHERE id=$4',
        "SUSPENDED", "REJECTED", now_utc(), worker_id,
    )
    user = await db.fetchrow('SELECT id, name FROM "User" WHERE id=$1', profile["userId"])
    return {"id": worker_id, "name": user["name"] if user else None, "status": "SUSPENDED"}
